Fix(jsx-classname-namespace): Accept root components in `main.jsx` files

bperson opened the pull request:
In the [devdocs's Hello World](https://wpcalypso.wordpress.com/devdocs/docs/guide/hello-world.md), we are currently breaking this ESLint rule, because the view-component is in a `main.jsx` file (to not conflict with `index.js` where we have the routing setup). A couple of others components around the calypso codebase also use `main.jsx` (37).

The main (limited) risk would be with directories containing both `index.jsx` and `main.jsx`. The only case currently in the calypso codebase is `/me/happychat`:

$ find . -type d -exec test -e {}/index.jsx -a -e {}/main.jsx \; -print
./me/happychat

By looking at the code, `index.jsx` is a controller-like calling `reactDom`, nothing fancy in it, making `main.jsx` the true container for a "root" component.

bperson committed:
Fix(jsx-classname-namespace): Accept root components in `main.jsx` files

aduth commented:
Begs a few questions:

- Regardless of precedent, is `main.jsx` a pattern we _want_ to encourage? How would we document this?
- Are there better ways we can identify the true root component of a folder without whitelisting a handful of acceptable file names?

bperson commented:
> Regardless of precedent, is main.jsx a pattern we want to encourage? How would we document this?

I think there is little value in enforcing/encouraging either pattern: they both provide value in different circumstances. The expected default exports of a "folder" can vary from setting up routing and/or an entry-point into rendering (use case for `index.js` + `main.jsx`) to a simple react component (`index.jsx`). Enforcing either would end up with more files and more boilerplate code in `index.js`, using it in most cases as a "proxy" for a folder's default exports.

That said, the current rule has led to `wp-super-cache__main` [for instance](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/wp-super-cache/main.jsx#L38) which defeats the purpose of accepting just the namespace on a root.

> Are there better ways we can identify the true root component of a folder without whitelisting a handful of acceptable file names?

Sadly, I don't think so as we're quickly hitting most linters' danger zone: multi-file. But it might be worth sleeping a few nights on it / doing further research [0]

[0] An intellectually-interesting idea (but unpractical + a lot of effort :o ) could be to run ESLint on webpack chunks, where each entire "folder" is contained inside a single file, parsable into a single AST.

aduth commented:
>That said, the current rule has led to `wp-super-cache__main` [for instance](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/wp-super-cache/main.jsx#L38) which defeats the purpose of accepting just the namespace on a root.

There's a part of me that thinks this is entirely reasonable and there shouldn't be such a thing as a "special root" component that lives alongside the router. For consistency's sake, why not treat it the same as any other component in that if you're seeking a specific class name, it _must_ exist as the `index.js` of the containing folder, even if that means `wp-super-cache/wp-super-cache/index.js` ?

It seems the `main.jsx` convention is pretty well solidified in Calypso so I'm not terribly opposed to this. Just trying to think outside the box to limit the number of consistency breaks and implied conventions we try to support.

bperson commented:
> There's a part of me that thinks this is entirely reasonable and there shouldn't be such a thing as a "special root" component that lives alongside the router. For consistency's sake, why not treat it the same as any other component in that if you're seeking a specific class name, it must exist as the index.js of the containing folder, even if that means wp-super-cache/wp-super-cache/index.js ?

I think there is value in having routing, rendering and the "root" component of a route as close to each others as possible as they share various parameters that tightly couple them together, introducing extra layers and folders will not really improve anything there.

That said, I don't feel strongly about it either :) But if we keep the current behaviour (enforcing usage of `index.jsx`), it's probably worth mentioning it in the rule's report message: I had to dig through the rule code to understand why the root component in hello world was infringing the rule and `wp-super-cache__main` is another example where the rule was misunderstood.

> It seems the main.jsx convention is pretty well solidified in Calypso so I'm not terribly opposed to this. Just trying to think outside the box to limit the number of consistency breaks and implied conventions we try to support.

And that's amazing, it's what makes code reviews worth it ;)

aduth commented:
One last question: Given that the convention is particularly Calypso-based and the rule could be useful outside of just the one project, do you think it makes sense to define as a setting to the rule, something like:

```js
"wpcalypso/jsx-classname-namespace": [ "error", {
	"rootFiles": [ "index.js", "index.jsx", "main.js", "main.jsx" ]
} ],
```

(Perhaps assigned as a regexp string)

See also: [`context.settings`](http://eslint.org/docs/developer-guide/working-with-rules#the-context-object)

Or similar to how `eslint-plugin-react` allows changes to the default `pragma` via top-level `settings.react`:

https://github.com/yannickcr/eslint-plugin-react#configuration

bperson committed:
Fix(jsx-classname-namespace): Add error message when everything but the filename is following our CSS guidelines

bperson committed:
Fix(jsx-classname-namespace): Add option `rootFiles` to overwrite default root files: (index|main).jsx?

bperson commented:
Updated:
- added a rule option: `rootFiles` [0]. I went the array route to not have to deal with escaping and modifiers when building a regex from string(s).
- added a more specific error message when everything is right ( classname + root component in the file ), but the filename doesn't match the configured `rootFiles`: 

`className should follow CSS namespace guidelines (expected root component to be in one of index.js, index.jsx)`

[0] As a sidenote: `options` seems to be rule-specific settings, whereas `settings` are higher-level settings at the eslint level.

aduth commented:
>[0] As a sidenote: `options` seems to be rule-specific settings, whereas `settings` are higher-level settings at the eslint level.

TIL 😃 

Few notes:

- There are a number of test failures. I'm guessing we'll at least need to update expected messaging around the changes here.
- Is there a better way we can determine whether the option has been assigned than to check the length? Seems hard to maintain if we include additional options later.
- Is it worth the effort to create a singular form of the expected warning message so it's less awkward a e.g. "to be in one of index.js."?

bperson committed:
Fix(jsx-classname-namespace): Cleaner options default

bperson committed:
Fix(jsx-classname-namespace): Add back support for filename with path components in them

bperson committed:
Fix(jsx-classname-namespace): Fix tests + add `options` tests and file error tests

bperson committed:
Fix(jsx-classname-namespace): Fix phrasing when using only one root file

bperson committed:
Fix(jsx-classname-namespace): Force `rootFiles` to be used with at least 1 item.

bperson commented:
> There are a number of test failures. I'm guessing we'll at least need to update expected messaging around the changes here.

All should be well now ;) 

Thanks to them, I also changed the message when we detect a className at the root of a file using the raw namespace instead of the prefix, the error message states more clearly the choice between "child"-ing the component via the prefix, or moving it to an accepted root file:

`className should follow CSS namespace guidelines (expected foo__ prefix or to be in one of index.js, index.jsx)`

> Is there a better way we can determine whether the option has been assigned than to check the length? Seems hard to maintain if we include additional options later.

Tried something, I'm not sure we are winning much but it's probably still something :) 

> Is it worth the effort to create a singular form of the expected warning message so it's less awkward a e.g. "to be in one of index.js."?

It's quite simple to do, so I threw a commit in there removing "one of" from the error message: 

`className should follow CSS namespace guidelines (expected foo__ prefix or to be in index.js`

aduth commented:
>Tried something, I'm not sure we are winning much but it's probably still something :)

Oh, I think I was confused and thinking that separate options would be defined at different indices in the array, but upon closer inspection I'm guessing defining options always follows the pattern of `[ "error", { } ]`

aduth commented on lib/rules/jsx-classname-namespace.js:
```diff
@@ -117,14 +113,15 @@ var rule = module.exports = function( context ) {
 
-	function isRootRenderedElement( node, filename ) {
+	function isFolderRootFile( filename ) {
+		const stripedFileName = filename.substr( filename.lastIndexOf( '/' ) + 1 );
```
Would `path.basename` work just as well here?

https://nodejs.org/api/path.html#path_path_basename_path_ext

Might at least indicate intent more clearly.

bperson replied:
fixed (and it's already used in the rule so you can add "consistency" to the list of reasons it's a better solution :) )

bperson committed:
Fix(jsx-classname-namespaces): Use `path.basename` to get the last part of filename

aduth reviewed (approved):
This looks good to me. I'll merge, tag, and release shortly.

aduth commented:
Just published v3.2.0

https://www.npmjs.com/package/eslint-plugin-wpcalypso

aduth commented:
Now that I think of it, we should probably update documentation for the rule to describe using the new option.

https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/jsx-classname-namespace.md

Do you think you could put together a pull request for that? We won't need to publish another release.

bperson commented:
Sure thing, should I also create a PR in [eslint-config-wpcalypso](https://github.com/Automattic/eslint-config-wpcalypso) setting the `rootFiles` config to accept `main.jsx?`?

aduth commented:
I think it might make most sense to specify that configuration in the Calypso repository itself, since it's very Calypso-specific (hence the idea for the option being extracted).
